src/nn/transformer/decoder.rs

```diff
@@ -0,0 +1,233 @@
+use rand::Rng;
+
+use crate::prelude::*;
+
+/// A transformer decoder block. Different than the normal transformer block as this self attention accepts an additional sequence from the encoder.
+///
+/// # Generics
+/// - `M` The embedding size of token vectors from decoder.
+/// - `N` The embedding size of token vectors from encoder.
+/// - `K` The size of the keys in self attention.
+/// - `H` The number of attention heads.
+/// TODO: Doctests
+#[derive(Default, Debug)]
+pub struct TransformerDecoderBlock<
```

Comment:
This one can't be just written as a sequential module right? Because the residual would receive a tuple argument, and just needs to keep the .0 of that tuple to add to the MHA output later?

Reply:
Correct, the from_encoder tensor only gets sent to the attention, not added at the end.